Dedup ModelChoiceField choices for limit_choices_to joins (swev-id: django__django-13315)

casey-brooks opened the pull request:
## Summary
- add regression models and tests covering ForeignKey choices filtered via joins in `limit_choices_to`
- deduplicate `ModelChoiceIterator` output by tracking `prepare_value()` keys while preserving the empty label and order
- ensure `ModelChoiceField.clean()` and `ModelMultipleChoiceField.clean()` tolerate duplicate join rows for the same object while surfacing `invalid_choice` for genuinely ambiguous to_field values

## Issue
- Closes #235

## Reproduction Steps
1. Checkout the base branch with the regression tests applied
2. PYTHONPATH=$PWD python tests/runtests.py model_forms.test_modelchoicefield.ModelChoiceFieldLimitChoicesToJoinTests --parallel=1

### Observed Failure
```text
Creating test database for alias 'default'...
Testing against Django installed in '/workspace/django/django'
System check identified no issues (0 silenced).
EFFF
======================================================================
ERROR: test_clean_accepts_selected_pk_after_dedup (model_forms.test_modelchoicefield.ModelChoiceFieldLimitChoicesToJoinTests.test_clean_accepts_selected_pk_after_dedup)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/workspace/django/tests/model_forms/test_modelchoicefield.py", line 409, in test_clean_accepts_selected_pk_after_dedup
    self.assertEqual(field.clean(str(self.alice.pk)), self.alice)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/django/django/forms/fields.py", line 149, in clean
    value = self.to_python(value)
            ^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/django/django/forms/models.py", line 1279, in to_python
    value = self.queryset.get(**{key: value})
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/django/django/db/models/query.py", line 439, in get
    raise self.model.MultipleObjectsReturned(
model_forms.models.Staff.MultipleObjectsReturned: get() returned more than one Staff -- it returned 2!

======================================================================
FAIL: test_limit_choices_to_join_deduplicates_choices (model_forms.test_modelchoicefield.ModelChoiceFieldLimitChoicesToJoinTests.test_limit_choices_to_join_deduplicates_choices)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/workspace/django/tests/model_forms/test_modelchoicefield.py", line 397, in test_limit_choices_to_join_deduplicates_choices
    self.assertEqual(choices[1:], [
AssertionError: Lists differ: [(<django.forms.models.ModelChoiceIteratorV[204 chars]no')] != [(1, 'Alice'), (2, 'Bruno')]

First differing element 1:
(<django.forms.models.ModelChoiceIteratorV[35 chars]ice')
(2, 'Bruno')

First list contains 1 additional elements.
First extra element 2:
(<django.forms.models.ModelChoiceIteratorValue object at 0xffff819ab1d0>, 'Bruno')

+ [(1, 'Alice'), (2, 'Bruno')]
- [(<django.forms.models.ModelChoiceIteratorValue object at 0xffff819ab050>,
-   'Alice'),
-  (<django.forms.models.ModelChoiceIteratorValue object at 0xffff819ab110>,
-   'Alice'),
-  (<django.forms.models.ModelChoiceIteratorValue object at 0xffff819ab1d0>,
-   'Bruno')]

======================================================================
FAIL: test_modelmultiplechoicefield_dedup_parity (model_forms.test_modelchoicefield.ModelChoiceFieldLimitChoicesToJoinTests.test_modelmultiplechoicefield_dedup_parity)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/workspace/django/tests/model_forms/test_modelchoicefield.py", line 428, in test_modelmultiplechoicefield_dedup_parity
    self.assertEqual(list(field.choices), [
AssertionError: Lists differ: [(<django.forms.models.ModelChoiceIteratorV[204 chars]no')] != [(1, 'Alice'), (2, 'Bruno')]

======================================================================
FAIL: test_to_field_name_dedup_and_clean (model_forms.test_modelchoicefield.ModelChoiceFieldLimitChoicesToJoinTests.test_to_field_name_dedup_and_clean)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/workspace/django/tests/model_forms/test_modelchoicefield.py", line 418, in test_to_field_name_deduplicates_choices
    self.assertEqual(choices[1:], [
AssertionError: Lists differ: [(<django.forms.models.ModelChoiceIteratorV[204 chars]no')] != [('ALICE', 'Alice'), ('BRUNO', 'Bruno')]

----------------------------------------------------------------------
Ran 4 tests in 0.004s

FAILED (failures=3, errors=1)
Destroying test database for alias 'default'...
```

## Testing
- PYTHONPATH=$PWD python tests/runtests.py model_forms.test_modelchoicefield --parallel=1
- flake8 django/forms/models.py tests/model_forms/models.py tests/model_forms/test_modelchoicefield.py

Python-side dedup preserves validation behaviour and avoids adding `distinct()` or other queryset changes.

casey-brooks committed:
fix(forms): dedupe foreign key choices

casey-brooks commented:
## Test & Lint Summary
- PYTHONPATH=/workspace/django python tests/runtests.py model_forms.test_modelchoicefield --parallel=1  *(27 tests, 0 failed, 0 skipped)*
- flake8 django/forms/models.py tests/model_forms/models.py tests/model_forms/test_modelchoicefield.py  *(no issues)*

noa-lucent commented on django/forms/models.py:
```diff
@@ -1272,11 +1278,15 @@ def prepare_value(self, value):
             value = self.queryset.get(**{key: value})
+        except MultipleObjectsReturned:
+            value = self.queryset.filter(**{key: value}).first()
+            if value is None:
```
[major] Catching  and returning  silently collapses any legitimate ambiguity in the queryset. In the join regression we expect duplicates to all refer to the *same* base row, but with this code we also accept unrelated duplicates (e.g. two different rows sharing the same ), and the form will now succeed while pointing at an arbitrary object. Previously this raised, making the data issue visible. Please keep the validation semantics strict: after the initial  fails, re-evaluate and ensure that all matches resolve to a single primary key (or raise ).

noa-lucent commented on django/forms/models.py:
```diff
@@ -1370,6 +1380,18 @@ def _check_values(self, value):
+            for obj in qs._result_cache:
+                identifier = str(prepare_value(obj))
+                if identifier in seen:
+                    continue
```
[major] This dedup step prunes duplicates purely by the serialized value. When  points at a non-unique column (allowed, albeit discouraged), this will now drop every extra row silently and the field will accept one arbitrary object instead of raising. Previously  returned the full queryset and the ambiguity was still visible. Please collapse only the duplicate rows that resolve to the *same* model instance (e.g. identical primary keys) and raise  if we encounter different objects with the same submitted value.

noa-lucent reviewed (changes requested):
Thanks for tackling the duplicate-choice regression. The iterator-side deduplication looks good, as do the regression models/tests. Unfortunately two blockers remain: catching  and calling  now masks genuinely ambiguous lookups instead of failing loudly, and the multiple-choice dedup block similarly drops distinct objects that share the serialized value. Please only collapse rows when they resolve to the same model instance and surface a validation error when they do not. Also, the PR body is missing an explicit link to issue #235.

casey-brooks committed:
fix(forms): refine dedup validation semantics

casey-brooks commented:
Addressed feedback:
- Restored `ModelChoiceField.to_python()` semantics to only tolerate duplicate join rows when they resolve to a single object (same PK) and raise `invalid_choice` for genuinely ambiguous to_field values.
- Trimmed `ModelMultipleChoiceField._check_values()` post-processing to deduplicate by primary key only, preserving distinct rows that share a serialized value.
- Added regression coverage ensuring non-unique `to_field_name` inputs raise a validation error instead of silently choosing an object.

Tests:
- PYTHONPATH=$PWD python tests/runtests.py model_forms.test_modelchoicefield --parallel=1
- flake8 django/forms/models.py tests/model_forms/models.py tests/model_forms/test_modelchoicefield.py

noa-lucent reviewed (approved):
Thanks for the fast follow-up. The iterator dedup now only skips repeated PKs,  raises  for genuinely ambiguous to_field values, and  keeps parity without collapsing distinct rows. The new non-unique to_field regression test fails pre-fix and passes now, and the PR body links #235. LGTM.
